packagekit: Add OS package update support

martinpitt opened the pull request:
Add a "Software Updates" page for classic dnf/yum/apt based systems,
to complement what "ostree" provides on Atomic. This uses the PackageKit
API for platform independence.

Set priority 0 to "hide" behind ostree if it is installed (which has
default priority 1), so that we don't try to mess around with packages
on OSTree/Atomic. Ship it in a new cockpit-packagekit rpm/deb as we 
don't want to pull in the packagekit dependency in cockpit-system (at 
least just yet).

Don't ship this on Ubuntu 16.04, as a PackageKit crash there
(https://launchpad.net/bugs/1689820) makes this functionality mostly
useless.

Specification:
https://github.com/cockpit-project/cockpit/wiki/Feature:-System-Updates-for-dnf,-yum,-apt-hosts

---
This is work in progress mostly due to missing tests. This has existed as a [standalone project](https://github.com/martinpitt/cockpit-os-updates) and got a lot of fixes and some third-party testing already, so it now becomes time to integrate it properly. I'd appreciate an initial coarse-grained review to point out code style and similar structural flaws, as this is by and large my first piece of React work.

 - [x] fix tooltips to be Cockpit-styled (semi-transparent, arrow pointing to widget)
 - [x] get PackageKit on all test images where missing (#6736)
 - [x] work around PackageKit online detection bug on debian-stable and ubuntu-1604 ([#1694438](https://bugs.launchpad.net/ubuntu/+source/packagekit/+bug/1694438))
 - [x] don't ship package updater on ubuntu-1604 due to [PackageKit crash](https://bugs.launchpad.net/ubuntu/+source/packagekit/+bug/1689820)
 - [x] create local rpm repository for yum/dnf with fake updates
 - [x] create local deb repository for  apt with fake updates
 - [x] add basic integration tests
 - [x] add integration tests for changelog parsing, with locally created updateinfo (yum/dnf)
 - [x] add integration tests for changelog parsing, with local changelog server (apt)
 - [x] add integration tests for installing security updates only
 - [x] split out into a separate binary package with a packagekit dependency
 - [x] fix tests on debian-stable
 - [x] newly separated binary package now requires PackageKit on selenium tests too, update fedora-24 image (PR #6850)
 - [x] merge binary packages from same source into the same table line?
 - [x] add progress bar for loading package indexes (checking for updates)
 - [x] fix tooltip pixel gap
 - [x] show if PK/queue is locked
 - [x] move applying updates from `componentDidMount()` to button action
 - [x] if applying updates fails, keep that error state and don't go into "reload updates" state
 - [x] resume upgrade transaction after reload/reconnect
 - [x] warn about disconnect if cockpit-ws is about tso be upgraded
 - [x] record a [video for this feature](https://youtu.be/BENWI-dTKMs)
 - [x] add a feature page
 - [x] rename to cockpit-packagekit

Shouldn't block this PR, can be a followup:
 - [ ] finalize design of "applying updates" page with @garrett (header bar vs `<h2>`, the rest is implemented)
 - [ ] implement new "applying updates" design
 - [ ] add a button to go back from error page to update list?

martinpitt commented:
Pushed initial basic tests, to see how far they get on all our images. Some of them don't have PackageKit installed and thus will have to be updated, and there might be other OS/release specific issues that I haven't considered/encountered yet.

martinpitt commented:
@garrett: The "installing updates" screen is currently a bit bare-bone, but the only reliable pieces of information that we have are the currently processed package, and the overall percentage:

![install](https://cloud.githubusercontent.com/assets/200109/26673480/56f4ed3e-46bd-11e7-89f2-eb52cfaee7b1.png)

Is this okay for the first version, or do you have some suggestions about the design/layout here?

garrett commented:
It's a bit tall, but here's an annotated mockup that covers the update steps.

![updates packages progress](https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/software-updates/software-updates-packages-progress.png)

martinpitt commented:
![](https://github.trello.services/images/mini-trello-icon.png) [OS package updates: interactive](https://trello.com/c/EmXDMIL8/499-os-package-updates%3A-interactive)

martinpitt commented:
![](https://github.trello.services/images/mini-trello-icon.png) [OS package updates: interactive](https://trello.com/c/EmXDMIL8/499-os-package-updates%3A-interactive)

martinpitt commented:
@garrett: I implemented some parts of the new design: the width-limited progress bar, no inline label any more, spelled out action label (Updating/Verifying/Downloading etc.), showing version number. *if* packagekit gives a time estimate, it is now shown on the right hand side too, but as dnf doesn't, it's not currently visible:

![install](https://cloud.githubusercontent.com/assets/200109/26715557/6b92b8fc-4776-11e7-9713-33ec53bac1cb.png)

Some issues I'd still like to discuss with you:

 - Using a header bar in some states (available updates, error loading/applying updates) but not during install seems a bit inconsistent. The above screenshot still keeps the current implementation of putting "Applying updates" into the header bar instead.
 - We don't show the host name on other Pages, just in the nagivation bar; it is often called "localhost" which isn't really pretty. Should we just leave that out?

martinpitt commented:
Splitting out the code into a new cockpit-pkgupdates binary package now requires adding PackageKit to at least Fedora-24 as well, for the selenium tests. I'll wait for the other test results and will then send a separate PR for that.

mvollmer commented:
The commit message has a duplicated paragraph "Set priority to 0 ...".

martinpitt commented:
@mvoller: Indeed, thanks! fixed locally, but won't push yet to avoid unnecessary CI reruns just for that.

mvollmer commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,448 @@
+    }
+
+    componentDidMount() {
+        cockpit.http(80, {address: 'cve.circl.lu'}).get('/api/cve/' + this.props.cve)
```
This will make a HTTP request from the cockpit-bridge process.  I think you want to use $.get here or similar, so that the browser makes that request.

martinpitt replied:
I'm hitting a wall here with `$.get()`. Adding cve.circl.lu to `content-security-policy` doesn't work. JSONP is [broken on cve.circl.lu](https://github.com/adulau/cve-search/issues/138), but even if it was working I'm not sure how that would help much.

Also, @petervo pointed out that this is a potential privacy issue, as cockpit is calling out to an external service without user confirmation. We decided to drop the popups and just keep them as simple links.

martinpitt replied:
Fixed now. I push FIXUPs for the time being instead of force-pushing, for easier incremental review. I'll squash them all together when this is ready to land.

mvollmer commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,448 @@
+    }
+
+    componentDidMount() {
+        pkTransaction()
```
It is surprising to see the actual action being initiated in a `componentDidMount` function.  This is pretty indirect and fragile, no?

martinpitt replied:
I'm open to suggestions where else  to put it. I can make it more indirect by moving it into a separate method and calling that from `componentDidMount()`, but the behaviour is exactly the same. What would be a better way?

mvollmer replied:
Something like this: https://github.com/mvollmer/cockpit/commit/ac75e2ba67fe5799638b3e61a4f12926a0204db4

With this, clicking the button directly starts the transaction, and ApplyUpdates is only watching it.

I believe your approach is also correct, but one needs to understand the React component lifecycle to see how the button click leads to the wanted transaction.  That's okay reading state from the server that is needed to render a component, but it's too indirect for user initiated actions that actually modify the server.

mvollmer commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,448 @@
+                });
+
+                $(transProxy).on('ErrorCode', (event, code, details) => {
+                    this.state.ErrorMessages.push(details);
```
`errorMessages`

martinpitt replied:
Fixed.

mvollmer commented:
From @abbra: It would be nice if we list source packages instead of binary packages.  The changelogs will be the same for all binary packages coming from the same source package.

For example, there is no real point in repeating the changelog info four times in this case:

![screenshot from 2017-06-02 14-21-33](https://cloud.githubusercontent.com/assets/3228183/26723782/ce088598-479e-11e7-986d-1dfd32cdab23.png)

martinpitt commented:
> It would be nice if we list source packages instead of binary packages

Indeed, these might be grouped into the same line, with the first column listing all binary packages. @garrett, WDYT? I made a TODO item in the checklist above.

martinpitt commented:
The PackageKit API doesn't actually expose source packages; but this can be done heuristically: if the version and changelog for two different binary packages are identical, then we can bucket them into the same table row and list all binary package names in the first cell.

martinpitt commented:
I now implemented this merging. It still needs corresponding test cases and it breaks the existing tests, but this is how it looks like:

![updates](https://cloud.githubusercontent.com/assets/200109/26737249/9740ea0c-47c9-11e7-8b44-b45de3c853c0.png)

@garrett might have an opinion how to prettify this?

martinpitt commented:
@mvollmer: This now addresses all your comments except for "initiated in componentDidMount", where I posted a question.

martinpitt commented:
Pushed again to fix a race in the tests, as the binary package names in the first table column weren't sorted.

mvollmer commented:
The tooltip sometimes has a gap between the box and the arrow.

![tooltip](https://cloud.githubusercontent.com/assets/3228183/26780094/ca242794-49f0-11e7-838f-25be7ef4c59f.png)

There is some CSS in pkg/storaged to fix it, maybe it is time to promote the fix to the tooltip itself.

```
/* Layout fix for tooltip arrows.
 *
 * Sometimes when there are non-integer coordinates involved
 * somewhere, rounding seems to make it so that there is a visible one
 * pixel gap between tooltip-inner and tooltip-arrow.  We avoid that
 * gap by letting the arrow and inner parts overlap by one pixel.
 *
 * This happens within the caption of the content listing.
 */

.tooltip.top .tooltip-arrow {
    bottom: 1px;
}

.tooltip.bottom .tooltip-arrow {
    top: 1px;
}
```

mvollmer commented:
> The PackageKit API doesn't actually expose source packages; but this can be done heuristically: if the version and changelog for two different binary packages are identical, then we can bucket them into the same table row and list all binary package names in the first cell.

Very nice!

mvollmer commented:
"Check for updates" should be more prominent, I'd say.  Right now it just shows "Loading available updates" as if the page is initializing itself with data from the server, but it does in fact goes non-trivial downloads from the repositories.  I think it should get it's own progress bar and cancel button.

mvollmer commented:
PackageKit has the PK_STATUS_ENUM_WAIT and PK_STATUS_ENUM_WAITING_FOR_LOCK state which i think we should show explicitly, and also for the GetUpdates transaction.

I had some really slow RefreshCache running in the background that I wasn't aware of, and the page just seemed dead.

I would argue that GetUpdates should never wait for a lock or in the queue, but it apparently does.

(I need to do the same for the Applications tool.)

mvollmer commented:
More from @abbra: On RHEL we should link to https://access.redhat.com/security/cve/ for CVEs.

Should we link to https://security-tracker.debian.org/tracker/ on Debian?  What about Ubuntu?

martinpitt commented:
> On RHEL we should link to https://access.redhat.com/security/cve/ for CVEs.

PackageKit gets the CVE URLs from the advisories. I wouldn't second-guess them, but use them as they are. If such a change is desirable, it should be changed in the actual advisories (but I suppose they would already use that URL). The fallback scanning is only for platforms which don't provide native CVE URLs in advisories (Fedora mostly),  and in that case I would keep the canonical mitre URL, as that is the institution that owns the CVE numbers.

martinpitt commented:
Notes wrt. locking: both `GetUpdates()` and `RefreshCache()` wait for exclusive access; during that time, at least on Fedora PK only sends a `PK_STATUS_ENUM_WAIT` update without any further progress updates until the lock gets released; I never saw `PK_STATUS_ENUM_WAITING_FOR_LOCK`. But `PK_STATUS_ENUM_WAIT` is always being sent, even if there is no other process locking the package database. Thus to avoid flicker, I propose to switch to a specific lock state ("Some other program is currently using the package manager, please wait...") after one second of receiving `WAIT` if there is no other status update after that.

martinpitt commented:
@mvollmer: I squashed the previous FIXUPs (that you already reviewed) into the main commit, and pushed four further FIXUPs that address your review comments. Note that I don't see the pixel gaps in the tooltips, but I copied your proposed CSS and it still looks alright to me.

martinpitt commented:
As @garrett is on holidays this week, I don't believe that the possible move from "Applying updates" from the header bar to a `<h2>` needs to block this PR, this can become a (rather simple) followup. There will be further follow-ups to this anyway, for "needs reboot" detection and configuring automatic updates.

mvollmer commented:
>> On RHEL we should link to https://access.redhat.com/security/cve/ for CVEs.

> PackageKit gets the CVE URLs from the advisories. I wouldn't second-guess them, but use them as they are. If such a change is desirable, it should be changed in the actual advisories (but I suppose they would already use that URL).

I see, thanks!

mvollmer commented:
I was able to see an error:

![screenshot from 2017-06-06 16-43-54](https://cloud.githubusercontent.com/assets/3228183/26832199/649b7354-4ad7-11e7-81a1-b5a271a8ad71.png)

The second box is completely empty, and shouldn't be shown, I guess.

mvollmer commented:
I got an update for `cockpit-ws` etc and during applying it, I got disconnected.  Should we do something about this?  Warn?  Fix cockpit not to disconnect during an update?

martinpitt commented:
I used that "stacked" layout as there are potentially multiple error messages during upgrade (multiple packages might fail), and I don't want to show only the last one as that is not necessarily the most useful one. Your first error message is from `UpdatePackages`, the second empty one from `GetUpdates()` -- apparently that failed with giving an empty error message. I suppose this is hard to reproduce.. But what I should totally fix: if applying the upgrades failed (left any error messages), then it should *not* automatically go into the "reload updates" state, but just  stay in that "show errors" state.

> I got an update for cockpit-ws etc and during applying it, I got disconnected

Just to be clear, the upgrade should still have gone on and finished in the background, just not with displaying it in Cockpit? Or did PackageKit interrupt the upgrade when the client disconnected? Right now, if you reconnect it should just say that it's waiting for the lock, but it might actually be possible to scan the running transactions (`busctl call org.freedesktop.PackageKit /org/freedesktop/PackageKit org.freedesktop.PackageKit GetTransactionList`) and get their role (`PK_ROLE_ENUM_UPDATE_PACKAGES`) and start listening to these signals,  i. e. visually resume the upgrade process. I'll play around with this.

martinpitt commented:
> I suppose this is hard to reproduce..

Heh, it's actually not - reproduced in a synthetic test case (manually for now, but simple to automate):
![updateerror](https://user-images.githubusercontent.com/200109/26866401-394cba2a-4b61-11e7-82fd-34865a24dad8.png)

mvollmer commented:
> Just to be clear, the upgrade should still have gone on and finished in the background, just not with displaying it in Cockpit?

Yes, that is what happened.

>Right now, if you reconnect it should just say that it's waiting for the lock,

Yes, it did.

> but it might actually be possible to scan the running transactions (busctl call org.freedesktop.PackageKit /org/freedesktop/PackageKit org.freedesktop.PackageKit GetTransactionList) and get their role (PK_ROLE_ENUM_UPDATE_PACKAGES) and start listening to these signals, i. e. visually resume the upgrade process. I'll play around with this.

Good idea.  I want to do the same when installing applications, but I think transactions don't have enough information to figure out which packages are actually being installed.  But in this case, we can just show the progress of the current transaction, whatever it is.

mvollmer commented:
> But what I should totally fix: if applying the upgrades failed (left any error messages), then it should not automatically go into the "reload updates" state, but just stay in that "show errors" state.

I think it stays in "loadError" state.  When UpdatePackages fails, the failure is handled by OsUpdated.handleError, no?  That function put the page into the "loadError" state where it stays until it is manually reloaded.

I think failures from UpdatePackages should not be handled by OsUpdates.handleError.  Did I introduce that problem with my `OsUpdates.apply_updates` method?

martinpitt commented:
I fixed the error handling during upgrades, including a test case. (Not pushed yet)

> Just to be clear, the upgrade should still have gone on and finished in the background, just not with displaying it in Cockpit?

I recreated this scenario with a test package that has a `%post sleep 100` and disconnected Cockpit with `systemctl stop cockpit`. The PK transaction *did* go on in the background, and the package update eventually finished. Reconnecting Cockpit then went into "Some other program is currently using the package manager, please wait..." as expected. So it appears "safe", but not yet pretty. How about:

 * Check if the packages to be updated include cockpit-ws, and put up a warning that the disconnection will happen.
 * On reconnect, check pending transactions, see if an upgrade is going on, and go immediately into "Applying updates" mode to watch it.

Both of these can be covered with tests.

martinpitt commented:
Marius Vollmer [2017-06-07  0:55 -0700]:
> I think it stays in "loadError" state.  When UpdatePackages fails, the failure is handled by OsUpdated.handleError, no?  That function put the page into the "loadError" state where it stays until it is manually reloaded.
>
> I think failures from UpdatePackages should not be handled by OsUpdates.handleError.

Correct. That's what I fixed now, and I renamed it to `handleLoadError()` for clarity.

> Did I introduce that problem with my `OsUpdates.apply_updates` method?

No, that was my fault already :-)

martinpitt commented:
I pushed four more FIXUPs that address @mvollmer's latest feedback: Upgrade progress is now being resumed after reconnect  or reload, upgrade errors are displayed properly, and there is a warning if cockpit-ws gets updated:

![cockpitupdate](https://user-images.githubusercontent.com/200109/26918329-d72ecb5e-4c31-11e7-8b6e-049a9bccdf03.png)

martinpitt commented:
I recorded a video for this new feature: https://youtu.be/BENWI-dTKM

stefwalter commented:
One general comment is that the page is very empty when no updates are available. A follow up pull request might add an http://www.patternfly.org/pattern-library/communication/empty-state/ to the page?

What do you think @garrett?

stefwalter commented on tools/cockpit.spec:
```diff
@@ -93,6 +93,7 @@ Recommends: %{name}-docker = %{version}-%{release}
 Suggests: %{name}-pcp = %{version}-%{release}
 Suggests: %{name}-kubernetes = %{version}-%{release}
 Suggests: %{name}-selinux = %{version}-%{release}
+Suggests: %{name}-pkgupdates = %{version}-%{release}
```
Are we sure about this name? Do you think we should make it be cockpit-packagekit ... to follow along with cockpit-ostree ... and with the assumption that other ways of updating a system will show up?

martinpitt replied:
I don't like cockpit-packagekit TBH:
 - This is soon going to interact with dnf/yum/apt specific configuration for setting up automatic updates (as PackageKit doesn't cover that).
 - PackageKit is not solely intended for updates, but also for finding and installing new software. For example, the app install bits that  @mvollmer is working on is also going to use PK. PK is a means to an end, not a functionality description.
 - *If* we are going to add other mechanisms to actually update packages (like calling yum directly instead of PackageKit), these would go into *this* module, not a different one.

stefwalter replied:
Okay that makes sense. But still not an trivial decision.

Our packages typically have not been named after functionality but after technology. So that multiple things with competing or different technologies can cover the same functionality ... and thus share a package alias (the ```"name"``` field in manifest.json).

> If we are going to add other mechanisms to actually update packages (like calling yum directly instead of PackageKit), these would go into this module, not a different one.

So does cockpit-ostree fould into this package or naming structure somehow? Or how does that fit into this picture? It's the same functionality with a different implementation.

martinpitt replied:
Right, ostree's manifest name is "updates", so it should probably have been called "ostree-updates", and we could call this "packagekit-updates" or "pk-updates" for brevity. To compare, we wouldn't put the functionality of searching and installing new apps (what @mvollmer is working on) into ostree just because that's also going to work on ostree.

Naming packages entirely after the underlying interfaces that they use is tricky -- for example, lots of our code talks to systemd,  not only pkg/systemd.

So, I'm not sure about the package name for this, I just have a strong feeling that "cockpit-packagekit" is really misleading and ambiguous, and it should be something else. Meeting topic?

stefwalter replied:
Sure. Names a hard ... maybe writing a Feature page will help you figure that out. 

https://github.com/cockpit-project/cockpit/tree/master/doc/guide

The feature pages are meant to help admins imagine how Cockpit is interacting with their system, and to give them clues on how to perform similar actions from the command line.

Once you think such a page through, maybe that'll help with figuring out a name. The package name should have a similar intent. Point toward the thing that we're interacting with, aid in diagnosis, quickly understanding how we're talking to the system.

martinpitt replied:
For the scope of this PR, the CLI equivalent is little more than "pkcon refresh" and "pkcon update" (I'll create a page for it). It's going to grow with configuring automatic updates and reboot detection though, as these need backend specific (yum/dnf/apt) commands. But I suppose the main conflict here is "stories/workflow" (apply/configure operating system updates) vs. "backend" (packagekit/dnf/apt/yum).

stefwalter commented on Makefile.am:
```diff
@@ -152,6 +152,7 @@ WEBPACK_PACKAGES = \
 	networkmanager \
 	ostree \
 	pcp \
+	pkgupdates \
```
See comment below about the package name. Likely 'packagekit'?

stefwalter commented on pkg/pkgupdates/index.html:
```diff
@@ -0,0 +1,39 @@
+  <link href="../base1/cockpit.css" rel="stylesheet">
+  <link href="pkgupdates.css" rel="stylesheet">
+
+  <script src="../base1/jquery.js"></script>
```
Are you actually using jquery? If not, no need to include it.

martinpitt replied:
I do use it for setting D-Bus signal handlers with `$(myDbusProxy).on(...)`. The JSX already `require()s` it, but without this `<script>` here I get an Oops about "ReferenceError: jQuery is not defined".

stefwalter replied:
In that case, I would suggest just using ```.addEventListener()``` and removing all use of jQuery.

martinpitt replied:
Nice! Done. Worth updating http://cockpit-project.org/guide/latest/cockpit-dbus.html for that? (In a separate PR)

stefwalter commented on pkg/pkgupdates/index.html:
```diff
@@ -0,0 +1,39 @@
+  <meta charset="utf-8">
+
+  <link href="../base1/patternfly.css" type="text/css" rel="stylesheet">
+  <link href="../base1/cockpit.css" rel="stylesheet">
```
Are you actually using cockpit.css? Typically all the css can be webpack bundled into pkgupdates.css rather than using this legacy file.

martinpitt replied:
Yes I do, mostly the `table.listing-ct` bits.

stefwalter replied:
Then you should ```require("listing.less");``` and don't worry about including this CSS file. This is legacy.

martinpitt replied:
Thanks, that works nicely. Fixed.

stefwalter commented on pkg/pkgupdates/index.html:
```diff
@@ -0,0 +1,39 @@
+<html>
+
+<head>
+  <title translate>Software Updates</title>
```
Does this actually get translated at runtime?

martinpitt replied:
I'm not sure how to verify that (i. e. how to create a PO template and update the translations). kubernetes, ostree, and systemd do the same, so I copied it from that.

stefwalter replied:
I think you just need to set the window.title at runtime during start up of your react based app.

martinpitt replied:
We don't set `window.title` anywhere in Cockpit. It's not really necessary either, as even without it the window title on http://localhost:9090/updates is fine, it's apparently built from the manifest's tools→updates→label plus the hostname. Do you know if setting it explicitly is useful for anything? Perhaps embedding as iframe in third-party pages or so? As every other pkg/* has a `<title>` I'm a bit hesitant to drop it.

stefwalter replied:
You don't need to drop the title tag. Just make sure it gets translated. It's marked for translation ... just make sure it gets translated in your app.

For example:

```
window.title = cockpit.gettext(window.title)
```

stefwalter replied:
This is the only translatable string outside of your React app. It makes sense to handle it properly.

Other apps that put everything in their HTML, use something like:

```
cockpit.translate()
```

to find and translate all these strings. I don't think you need that for this one string.

martinpitt replied:
That makes perfect sense, thanks for the explanation. Fixed.

stefwalter commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,532 @@
+var $ = require('jquery');
```
Should this file have a license header? Or is that unnecessary?

martinpitt replied:
Yes it should; added (not pushed yet, will do once I addressed all your feedback).

stefwalter commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,532 @@
+
+// 'available' heading is built dynamically
+const STATE_HEADINGS = {
+    'loading': _('Loading available updates, please wait...'),
```
All translatable strings need to use double quotes. This is due to the gettext C parser. For that reason it's a good habit to make all strings use double quotes.

Either that, or you could work on a different parser for javascript. ;)

martinpitt replied:
```
$ git grep "'" pkg/ | wc -l; git grep '"' pkg |  wc -l
4856
29498
```
OK, the double quotes have it by majority :-) Changed.

stefwalter commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,532 @@
+            if (props.updates[u].security)
+                ++num_security;
+        if (num_security > 0)
+            state += cockpit.ngettext(', including $1 security fix', ', including $1 security fixes',  num_security);
```
Unfortunately, this is not a localizable way of assembling a string. It's likely these will need to be multiple sentences rather than a bunch of translatable clauses that are combined.

martinpitt replied:
Since there is no multi-dimensional ngettext (the combinatorial explosion would kill translators), we need to keep the two apart somehow. We could use parentheses or a slash as a separator instead of a comma in the C string, but this is not really a sentence but an enumeration.

if* there is a language that doesn't have comma enumeration and the order of "a, including b" is reversed with no logical possibility of using the English order, then that translation should reformat it as appropriate - it could say  "$0 qu'moH" and " ($1 niH ga moH)", anything which keeps the security number second.

stefwalter commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,532 @@
+            state += cockpit.ngettext(', including $1 security fix', ', including $1 security fixes',  num_security);
+        state = cockpit.format(state, num_updates, num_security);
+    } else
+        state = STATE_HEADINGS[props.state];
```
Lets use braces around if/else blocks if any part of the if/else is multiline.

martinpitt replied:
Fixed.

stefwalter commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,532 @@
+                <tr>
+                    <td id='state'>{state}</td>
+                    {props.state == 'uptodate' || props.state == 'available' ?
+                        <td className='text-right'>{props.timeSinceRefresh ?
```
The use of trinary inline if operator here makes this hard to maintain. Is there another solution that could be used?

martinpitt replied:
I find it easier to read and understand the structure like that, but I don't have a strong opinion. Pulled out into variables.

stefwalter commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,532 @@
+
+    if (info.bug_urls && info.bug_urls.length)
+        // we assume a bug URL ends with a number; if not, show the complete URL
+        bugs = commaJoin(info.bug_urls.map(u => <a rel='noopener' referrerpolicy='no-referrer' target='_blank' href={u}>{u.match(/[0-9]+$/) || u}</a>));
```
Multiline if should have braces.

martinpitt replied:
Fixed.

stefwalter commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,532 @@
+                </div>
+
+                {this.state.percentage !== null
+                    ? (<div className='progress progress-label-top-right'>
```
Use of trinary here makes this hard to maintain.

martinpitt replied:
Pulled out into a variable, similarly to above.

stefwalter commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,532 @@
+                    }
+                });
+
+
```
Extra line.

martinpitt replied:
fixed

stefwalter commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,532 @@
+        // return type is 'u', but returns -1 if not supported; so ignore implausibly high values (> 10 years)
+        dbus_pk.call('/org/freedesktop/PackageKit', 'org.freedesktop.PackageKit', 'GetTimeSinceAction',
+                     [PK_ROLE_ENUM_REFRESH_CACHE], {timeout: 5000})
+            .done(seconds => (seconds > 315360000 || this.setState({timeSinceRefresh: seconds})))
```
This magic number here could use a constant.

martinpitt replied:
Fixed.

stefwalter commented on test/verify/check-pkgupdates:
```diff
@@ -0,0 +1,445 @@
+from testlib import *
+
+@skipImage("Image uses OSTree", "continuous-atomic", "fedora-atomic", "rhel-atomic")
+@skipImage("PackageKit crashes, https://launchpad.net/bugs/1689820", "ubuntu-1604")
```
This should be a known issue, not a skip.

martinpitt replied:
It needs to be a skip as currently cockpit-pkgupdates isn't built at all for ubuntu-1604 due to this bug. (See below).

stefwalter replied:
It can't build or just isn't building right now? We should build it even if an operating system bug prevents it from working ... and track it as a known issue.

martinpitt replied:
It could build, but I added blacklisting to tools/debian/rules. I really disagree about shipping it in a known completely broken state. I'm fine with building it on ubuntu-1604 for our CI, but not on "the real thing" (on releases and in Ubuntu xenial-backports).

stefwalter commented on test/verify/check-pkgupdates:
```diff
@@ -0,0 +1,445 @@
+        if self.machine.image in ["ubuntu-1604", "debian-stable"]:
+            # old PackageKit+NM on Debian/Ubuntu misdetect online status with ifupdown; work around
+            # https://launchpad.net/bugs/1694438
+            # TODO: use a naughty override instead?
```
Naughty override if it affects real users using the feature in Cockpit.

martinpitt replied:
Can do, but this means that we wouldn't test OS updates on debian-stable *at all*. It wouldn't affect debian-stable users who use ifupdown (i. e. servers), or use NetworkManager without a static ifupdown config (i. e. desktops). This just affects the case of ifupdown *plus* NM that we use on our test images. So I thought it would be more useful to run the checks with a config that works than to skip them entirely due to our special setup.

stefwalter replied:
Makes sense, if it doesn't affect most users, it's fine to work around it.

stefwalter commented on tools/debian/rules:
```diff
@@ -8,6 +8,12 @@ ifneq ($(shell dpkg -s libpcp3-dev >/dev/null 2>&1 && echo yes),yes)
 endif
 
+# PackageKit crashes on update information on Ubuntu 16.04, which makes
+# "Software Updates" useless (LP: #1689820)
```
This is not a reason not to ship the feature. That's a bug in Ubuntu. And we should track it as a known naughty issue.

martinpitt replied:
Sure it is a bug in Ubuntu, and I filed it properly downstream. But this is a dealbreaker - I didn't find a workaround, and not being able to get updates makes this page entirely useless. I am not going to ship a completely broken new package, we can enable it later if/once the bug gets fixed. Note that pkgupdates does get shipped on ubuntu 17.04, just not on 16.04.

Some background: For a long time Ubuntu has used aptdaemon as a PackageKit compatible (-ish) implementation as PackageKit's apt support was rather poor in the past. In newer versions PackageKit itself has gotten much more attention, and Ubuntu moves away from aptdaemon.

stefwalter commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,532 @@
+    loadUpdateDetails(pkg_ids) {
+        pkTransaction()
+            .done(transProxy => {
+                $(transProxy).on('UpdateDetail', (event, packageId, updates, obsoletes, vendor_urls,
```
This should just use transProxy.addEventListener('UpdateDetail', ... )

stefwalter commented:
This also needs a feature page:

https://github.com/cockpit-project/cockpit/tree/master/doc/guide

martinpitt commented:
I pushed the FIXUPs for @stefwalter's review comments, except for "add feature page" and "discuss package name" which I added to the TODO list in the description for now.

martinpitt commented:
Feature page pushed too now, CI was happy with the previous FIXUPs.

mvollmer commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,558 @@
+
+            case "loadError":
+            case "updateError":
+                return this.state.errorMessages.map(m => <pre>{m}</pre>);
```
There is no way out of this state except by reloading the page.  I think there should be a way to go back to 'loading'.

martinpitt replied:
I'm not sure - if the upgrade fails, you should really investigate on the command line, as there is no interactivity in the PackageKit upgrade. The user might try to re-run the update, but that might make things worse (or at least not better). I'm okay with adding some "Acknowledge" or "Go back" button there in case a retry does help. @garrett, do you have some input on that?

I added it as a TODO item for now, but I don't think this necessarily needs to block this PR.

mvollmer replied:
> I added it as a TODO item for now, but I don't think this necessarily needs to block this PR.

I agree.

mvollmer commented:
For some reason, the table isn't full-width anymore.  I couldn't quickly figure out what has changed.

![screenshot from 2017-06-12 12-38-11](https://user-images.githubusercontent.com/3228183/27028001-1b74d320-4f6c-11e7-9f92-c0b8ab6dd604.png)

martinpitt commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,558 @@
+const PK_INFO_ENUM_SECURITY = 8;
+const PK_STATUS_ENUM_WAIT = 1;
+const PK_STATUS_ENUM_UPDATE = 10;
+const PK_STATUS_ENUM_WAITING_FOR_LOCK = 1;
```
Fixed locally (didn't push yet), thanks!

martinpitt commented on pkg/pkgupdates/pkgupdates.jsx:
```diff
@@ -0,0 +1,558 @@
+const PK_STATUS_ENUM_WAITING_FOR_LOCK = 1;
+
+const PK_STATUS_STRINGS = {
+    9: _("Downloading"),
```
Dito, thanks!

mvollmer replied:
I have never seen the "Downloading" state, actually.  I tried to provoke it, but failed. Maybe this is a PackageKit quirk on Fedora.

martinpitt replied:
Yes, me neither, the download part is reported as `INSTALL` on Fedora. But I think it can't hurt to provide a human readable string for it anyway - it may get fixed in PackageKit one day, or already be used with other backends (yum, apt, etc.)

martinpitt commented:
> For some reason, the table isn't full-width anymore. I couldn't quickly figure out what has changed.

I figure what's changed is that this time you only have one update with a very short description, no bugs etc. On "design patterns" the listing-ct isn't full width either, and it would only make the table harder to read to add extra whitespace. I wouldn't like to divert from cockpit's standard listing-ct here. I have 40 updates (since Friday) and the table is full width, with line breaks.

mvollmer commented:
> I wouldn't like to divert from cockpit's standard listing-ct here. 

Makes sense, thanks.

martinpitt commented:
@mvollmer: In today's [weekly meeting](https://meetbot.fedoraproject.org/cockpit/2017-06-12/cockpit_weekly_meeting.2017-06-12-13.02.log.html) we discussed the package name again, between "cockpit-pkgupdates" and "cockpit-packagekit". Do you have some thoughts how this would relate to your upcoming "app discovery/install" page? That will use PackageKit, docker, and (on Atomic) possibly OSTree? Does it make sense for it to depend on "cockpit-packagekit" or even go into that same binary package? How about the module name in `pkg/`? @petervo and I would appreciate your thoughts here. Thanks!

garrett commented:
@stefwalter: I agree. I didn't design an empty state yet — but definitely should. Thanks for pointing that out! 👍

da2x commented:
cockpit-pkgupdates gets stuck on `Loading available updates, please wait...` forever. It should probably times out after a while and show and error. (`pkcon refresh` crashes on current f26.)

martinpitt commented:
@da2x: I can easily reproduce the [F26 PackageKit refresh crash](https://bugzilla.redhat.com/show_bug.cgi?id=1460825), that was introduced only recently. This corresponds to the progress bar getting stuck at some 20% indeed. This doesn't send out an `Error` signal (quite naturally, as PK crashed), so indeed this needs to be handled at the D-Bus connection level.

I now have a FIXUP with corresponding test case that handles PackageKit crashes. I will push it in a bit as I also change the "empty state" page according to @garrett 's designs and don't want to trigger CI too often.

garrett commented:
I made some empty state mockups, shared them with @martinpitt over IRC today, and we refined them iteratively (mainly for the current lack of auto-updates).

![noupdates](https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/software-updates/software-updates-packages-noupdates.png)

...and realized we should have a stale state, which automatically triggers a check for new packages if someone is visiting the page and there hasn't been a check in a while. (In the mockup, I have this at >= 1 day.)

![stale](https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/software-updates/software-updates-packages-stale.png)

The available packages header is a little awkward here — especially when there are no auto-updates, so @martinpitt and I decided to drop the header altogether in these states for now. We'll revisit the header (and the text it may contain... if it should be different) when the automatic updates feature is added in.

garrett commented:
In IRC, @martinpitt mentioned that the update progress is actually known, so I'll look into modifying these mockups. They'll probably change inline to the latest versions, as I hotlinked them from their master version in the design repo.

martinpitt commented:
I pushed four fixups now:
 - Fix wrong PK status enum values that @mvollmer spotted
 - Handle PackageKit crashes
 - Prettier "no updates" page (from @garrett)
 - Automatically trigger refresh when last checked ≥ 1 day ago (from @garrett above)

martinpitt commented:
Force-pushed again to rename from "pkgupdates" to "packagekit" as per team discussion. I also squashed all the fixups together.

martinpitt commented:
@mvollmer, @stefwalter : This is now all green and should address all your feedback.
